File: old_division/ARC/ARC106/b.cc
#include <iostream>
#include <vector>
#include <string>

using namespace std;

struct UnionFind {
    vector<int> par; // par[i]:iの親の番号　(例) par[3] = 2 : 3の親が2

    UnionFind(int N) : par(N) { //最初は全てが根であるとして初期化
        for(int i = 0; i < N; i++) par[i] = i;
    }

    int root(int x) { // データxが属する木の根を再帰で得る：root(x) = {xの木の根}
        if (par[x] == x) return x;
        return par[x] = root(par[x]);
    }

    void unite(int x, int y) { // xとyの木を併合
        int rx = root(x); //xの根をrx
        int ry = root(y); //yの根をry
        if (rx == ry) return; //xとyの根が同じ(=同じ木にある)時はそのまま
        par[rx] = ry; //xとyの根が同じでない(=同じ木にない)時：xの根rxをyの根ryにつける
    }

    bool same(int x, int y) { // 2つのデータx, yが属する木が同じならtrueを返す
        int rx = root(x);
        int ry = root(y);
        return rx == ry;
    }
};

int main() {
    int N, M;
    cin >> N>>M;
    int s;
    UnionFind tree(N);
    int a[N];
    int b[N];
    int as[N];
    int bs[N];
    for(int i = 0; i < N; i++) {
        cin >> a[i];
    }
    for(int i = 0; i < N; i++) {
        cin >> b[i];
    }
    int c,d;
    for(int i = 0; i < M; i++) {
        cin >> c >> d;
        tree.unite(c-1, d-1);
    }
    for(int i = 0; i < N; i++) {
        s=tree.root(i);
        as[s]+=a[s];
        bs[s]+=b[s];
    }
    for(int i = 0; i < N; i++) {
        if(as[i]!=bs[s]){
          cout << "No";
          return 0;
        }else{
          continue;
        }
    }
  cout << "Yes";

    return 0;
}

File: old_division/ARC/ARC106/b.py
n,m=map(int,input().split())
a=list(map(int,input().split()))
b=list(map(int,input().split()))
class UnionFind():
    def __init__(self, n):
        self.n = n
        self.parents = [-1] * n

    def find(self, x):
        if self.parents[x] < 0:
            return x
        else:
            self.parents[x] = self.find(self.parents[x])
            return self.parents[x]

    def union(self, x, y):
        x = self.find(x)
        y = self.find(y)

        if x == y:
            return

        if self.parents[x] > self.parents[y]:
            x, y = y, x

        self.parents[x] += self.parents[y]
        self.parents[y] = x

    def size(self, x):
        return -self.parents[self.find(x)]

    def same(self, x, y):
        return self.find(x) == self.find(y)

    def members(self, x):
        root = self.find(x)
        return [i for i in range(self.n) if self.find(i) == root]

    def roots(self):
        return [i for i, x in enumerate(self.parents) if x < 0]

    def group_count(self):
        return len(self.roots())

    def all_group_members(self):
        return {r: self.members(r) for r in self.roots()}

    def __str__(self):
        return '\n'.join('{}: {}'.format(r, self.members(r)) for r in self.roots())

uf=UnionFind(n)
for i in range(m):
    c,d=map(int,input().split())
    uf.union(c-1,d-1)

root=uf.roots()

h=[0]*n
d=[0]*n
for i in range(n):
    roo=uf.find(i)
    h[roo]+=a[roo]
    d[roo]+=b[roo]

for i in range(n):
    if h[i]!=d[i]:
        print("No")
        exit()
    else:
        continue


for i in root:
    member =uf.members(i)
    asum=0
    bsum=0
    for j in member:
        asum+=a[j]
        bsum+=b[j]
    if asum==bsum:
        continue
    else:
        print("No")
        exit()

print("Yes")

